### ISSUE TITLE
Bug: Gluetun thinks ipv6 is supported while disabled on host

### ISSUE THREAD

**Issue by maanoobh** [qdm12/gluetun#1241]
### Is this urgent?

No

### Host OS

Ubuntu 22.04

### CPU arch

x86_64

### VPN service provider

Custom

### What are you using to run the container

docker-compose

### What is the version of Gluetun

Running version latest built on 2022-11-11T09:55:48.270Z (commit 7421dcb)

### What's the problem 🤔

It seems that Gluetun thinks that ipv6 is supported when it’s disabled on the host, and trying to add routes for it and failing. 

### Share your logs

```plain text
========================================
=============== gluetun ================
========================================
=========== Made with ❤️ by ============
======= https://github.com/qdm12 =======
========================================
========================================

Running version latest built on 2022-11-11T09:55:48.270Z (commit 7421dcb)

🔧 Need help? https://github.com/qdm12/gluetun/discussions/new
🐛 Bug? https://github.com/qdm12/gluetun/issues/new
✨ New feature? https://github.com/qdm12/gluetun/issues/new
☕ Discussion? https://github.com/qdm12/gluetun/discussions/new
💻 Email? quentin.mcgaw@gmail.com
💰 Help me? https://www.paypal.me/qmcgaw https://github.com/sponsors/qdm12
2022-11-14T21:30:54+03:00 INFO [routing] default route found: interface eth0, gateway 172.27.0.1 and assigned IP 172.27.0.2
2022-11-14T21:30:54+03:00 INFO [routing] local ethernet link found: eth0
2022-11-14T21:30:54+03:00 INFO [routing] local ipnet found: 172.27.0.0/16
2022-11-14T21:30:54+03:00 INFO [firewall] enabling...
2022-11-14T21:30:54+03:00 INFO [firewall] enabled successfully
2022-11-14T21:30:55+03:00 INFO [storage] merging by most recent 13220 hardcoded servers and 13220 servers read from /gluetun/servers.json
2022-11-14T21:30:56+03:00 INFO Alpine version: 3.16.2
2022-11-14T21:30:56+03:00 INFO OpenVPN 2.4 version: 2.4.12
2022-11-14T21:30:56+03:00 INFO OpenVPN 2.5 version: 2.5.6
2022-11-14T21:30:56+03:00 INFO Unbound version: 1.15.0
2022-11-14T21:30:56+03:00 INFO IPtables version: v1.8.8
2022-11-14T21:30:56+03:00 INFO Settings summary:
├── VPN settings:
|   ├── VPN provider settings:
|   |   ├── Name: custom
|   |   └── Server selection settings:
|   |       ├── VPN type: wireguard
|   |       ├── Target IP address: 198.xxx.xxx.235
|   |       └── Wireguard selection settings:
|   |           ├── Endpoint IP address: 198.xxx.xxx.235
|   |           ├── Endpoint port: 35000
|   |           └── Server public key: DbbdOp9o9QO2IXNmGJfxJxyjAAajA+n9hwkVuts2rVg=
|   └── Wireguard settings:
|       ├── Private key: AE...Wo=
|       ├── Interface addresses:
|       |   └── 10.0.0.4/32
|       └── Network interface: tun0
├── DNS settings:
|   ├── DNS server address to use: 127.0.0.1
|   ├── Keep existing nameserver(s): no
|   └── DNS over TLS settings:
|       ├── Enabled: yes
|       ├── Update period: every 24h0m0s
|       ├── Unbound settings:
|       |   ├── Authoritative servers:
|       |   |   └── cloudflare
|       |   ├── Caching: yes
|       |   ├── IPv6: no
|       |   ├── Verbosity level: 1
|       |   ├── Verbosity details level: 0
|       |   ├── Validation log level: 0
|       |   ├── System user: root
|       |   └── Allowed networks:
|       |       ├── 0.0.0.0/0
|       |       └── ::/0
|       └── DNS filtering settings:
|           ├── Block malicious: yes
|           ├── Block ads: no
|           ├── Block surveillance: no
|           └── Blocked IP networks:
|               ├── 127.0.0.1/8
|               ├── 10.0.0.0/8
|               ├── 172.16.0.0/12
|               ├── 192.168.0.0/16
|               ├── 169.254.0.0/16
|               ├── ::1/128
|               ├── fc00::/7
|               ├── fe80::/10
|               ├── ::ffff:7f00:1/104
|               ├── ::ffff:a00:0/104
|               ├── ::ffff:a9fe:0/112
|               ├── ::ffff:ac10:0/108
|               └── ::ffff:c0a8:0/112
├── Firewall settings:
|   └── Enabled: yes
├── Log settings:
|   └── Log level: INFO
├── Health settings:
|   ├── Server listening address: 127.0.0.1:9999
|   ├── Target address: cloudflare.com:443
|   ├── Read header timeout: 100ms
|   ├── Read timeout: 500ms
|   └── VPN wait durations:
|       ├── Initial duration: 6s
|       └── Additional duration: 5s
├── Shadowsocks server settings:
|   └── Enabled: no
├── HTTP proxy settings:
|   └── Enabled: no
├── Control server settings:
|   ├── Listening address: :8000
|   └── Logging: yes
├── OS Alpine settings:
|   ├── Process UID: 1000
|   ├── Process GID: 1000
|   └── Timezone: Asia/Bahrain
├── Public IP settings:
|   ├── Fetching: every 12h0m0s
|   └── IP file path: /tmp/gluetun/ip
└── Version settings:
    └── Enabled: yes
2022-11-14T21:30:56+03:00 INFO IPv6 is supported
2022-11-14T21:30:56+03:00 INFO [routing] default route found: interface eth0, gateway 172.27.0.1 and assigned IP 172.27.0.2
2022-11-14T21:30:56+03:00 INFO [routing] adding route for 0.0.0.0/0
2022-11-14T21:30:56+03:00 INFO [firewall] setting allowed subnets...
2022-11-14T21:30:56+03:00 INFO [routing] default route found: interface eth0, gateway 172.27.0.1 and assigned IP 172.27.0.2
2022-11-14T21:30:56+03:00 INFO [pprof] http server listening on 0.0.0.0:6060
2022-11-14T21:30:56+03:00 INFO [dns over tls] using plaintext DNS at address 1.1.1.1
2022-11-14T21:30:56+03:00 INFO [http server] http server listening on 0.0.0.0:8000
2022-11-14T21:30:56+03:00 INFO [firewall] allowing VPN connection...
2022-11-14T21:30:56+03:00 INFO [healthcheck] listening on 127.0.0.1:9999
2022-11-14T21:30:56+03:00 INFO [wireguard] Using available kernelspace implementation
2022-11-14T21:30:56+03:00 INFO [wireguard] Connecting to 198.xxx.xxx.235:35000
2022-11-14T21:30:56+03:00 ERROR [vpn] setting up IPv6: cannot add route for interface: cannot add route for link tun0, destination ::/0 and table 51820: operation not supported
2022-11-14T21:30:56+03:00 INFO [vpn] retrying in 15s
2022-11-14T21:31:02+03:00 INFO [healthcheck] program has been unhealthy for 6s: restarting VPN
```

### Share your configuration

```yml
version: "2.1"
services:
  gluetun:
    image: qmcgaw/gluetun
    container_name: gluetun
    cap_add:
      - NET_ADMIN
    devices:
      - /dev/net/tun:/dev/net/tun
    volumes:
      - ./gluetun:/gluetun
    ports:
      - 9091:9091
    environment:
      - VPN_SERVICE_PROVIDER=custom
      - VPN_TYPE=wireguard
      - VPN_ENDPOINT_IP=198.xxx.xxx.235
      - VPN_ENDPOINT_PORT=35000
      - WIREGUARD_PUBLIC_KEY=DbbdO…
      - WIREGUARD_PRIVATE_KEY=AEU…
      - WIREGUARD_ADDRESSES=10.0.0.4/32
      - TZ=Asia/Bahrain
```

**Comment by Dared87** [comment]
Same here, I get similar errors because it detects ipv6, but it is disabled on host:
`2022-11-18T21:35:35Z INFO [openvpn] Linux ip -6 addr add failed: external program exited with error status: 2
2022-11-18T21:35:50Z INFO [openvpn] TCP/UDP: Preserving recently used remote address: [AF_INET6]xxxx:xxxx:x:xxxx::xxx:xxxx
2022-11-18T21:35:50Z INFO [openvpn] UDP: Cannot create UDP/UDP6 socket: Address family not supported by protocol (errno=97)`

This happens on version 3.32.0, 3.31.1 is still working.

**Comment by Makifun** [comment]
Just wanted to chime in and say same here, using wireguard. IPv6 disabled via grub in host, same error in the latest, v3.31.1, v3.30.1 and v3.29.0. Would be great if we could manually disable ipv6 in the container. :)

`2022-11-27T15:13:18+01:00 ERROR [vpn] setting up IPv6: cannot add route for interface: cannot add route for link tun0, destination ::/0 and table 51820: operation not supported`

**Comment by Dared87** [comment]
> IPv6 disabled via grub in host

I did the same btw.

**Comment by qdm12** [comment]
The problem is likely due to https://github.com/qdm12/gluetun/commit/5ddd703f6ab746c029cbd8b58466c8eae822bda5 or maybe https://github.com/qdm12/gluetun/commit/875690ab18eaea5d85739f9f620e273ed54c3185 (more unlikely).

If you do:

```sh
docker run -it --rm alpine:3.16 ip -6 route
```

Do you get any route? As far as I know, Gluetun assumes IPv6 is available as soon as it finds a route using an IPv6 address. I'll dig more in the code tomorrow morning, maybe push an image with debug logs if I can't solve it.

**Comment by qdm12** [comment]
- @Makifun it seems it only fails on v3.32.0 for @Dared87 maybe you have a different issue?
- @maanoobh does this fail only on latest and v3.32.0, or does it fail with v3.31.1 as well?
- I can have a setting to force IPv6 to be enabled or disabled, but I would rather keep it automated ideally, so I'm going to try resolving this :pray:

**Comment by Dared87** [comment]
I'm using podman so
`podman run -it --rm alpine:3.16 ip -6 route`
outputs
`default via 10.0.2.2 dev tap0 
10.0.2.0/24 dev tap0 scope link  src 10.0.2.100`

The same command returns nothing on the host machine (as expected). So this is not the fault of gluetun, it is mine. But good to know how to check for it. I'll have a look at how its possible to deactivate ipv6 inside podman.

Thank you for the reply!

**Comment by qdm12** [comment]
@Dared87 1b1335835bd9ac803281e9f7c9b79d4d6e33dec1 now inspects each route returned from the equivalent kernel call to `ip -6 route`, and only detects ipv6 as supported if at least one route has a destination or source IP which is IPv6. That should fix it for Podman without having to wait a few seconds.

@maanoobh @Makifun please run `docker run -it --rm alpine:3.16 ip -6 route` and let me know what it returns for you.

**Comment by Makifun** [comment]
@qdm12 Sure! Ran it once without specifying any network and it returned something(seems ipv4 related only though) and then I ran it while specifying my network(with enable_ipv6: false set in docker-compose) that all my containers run under and it returned nothing. Also ran it from within the gluetun container(latest image) just in case(it is connected to medializr_default). 🤔 

```
makifun@wololo:~$ docker run -it --rm alpine:3.16 ip -6 route
default via 172.17.0.1 dev eth0
172.17.0.0/16 dev eth0 scope link  src 172.17.0.2
makifun@wololo:~$ docker run -itd --network=medializr_default --rm alpine:3.16 ip -6 route
de22b8f7b16452afcd782297ef383c390e48ce2e329ca8ac2738649ae937e04d
makifun@wololo:~/medializr$ docker exec -it gluetun ip -6 route
makifun@wololo:~/medializr$ docker exec -it gluetun ip -4 route
default via 172.19.0.1 dev eth0
172.19.0.0/16 dev eth0 proto kernel scope link src 172.19.0.24
```

**Comment by Dared87** [comment]
@qdm12 Thank you for your support! I did a test, but got a runtime error

```
========================================
========================================
=============== gluetun ================
========================================
=========== Made with ❤️ by ============
======= https://github.com/qdm12 =======
========================================
========================================

Running version latest built on 2022-12-01T12:23:13.075Z (commit 1b13358)

🔧 Need help? https://github.com/qdm12/gluetun/discussions/new
🐛 Bug? https://github.com/qdm12/gluetun/issues/new
✨ New feature? https://github.com/qdm12/gluetun/issues/new
☕ Discussion? https://github.com/qdm12/gluetun/discussions/new
💻 Email? quentin.mcgaw@gmail.com
💰 Help me? https://www.paypal.me/qmcgaw https://github.com/sponsors/qdm12
2022-12-01T15:19:12Z INFO [routing] default route found: interface tap0, gateway 10.0.2.2 and assig>
2022-12-01T15:19:12Z INFO [routing] local ethernet link found: tap0
2022-12-01T15:19:12Z INFO [routing] local ipnet found: 10.0.2.0/24
2022-12-01T15:19:12Z INFO [firewall] enabling...
2022-12-01T15:19:13Z INFO [firewall] enabled successfully
2022-12-01T15:19:13Z INFO [storage] merging by most recent 13222 hardcoded servers and 13222 server>
2022-12-01T15:19:13Z INFO Alpine version: 3.16.3
2022-12-01T15:19:13Z INFO OpenVPN 2.4 version: 2.4.12
2022-12-01T15:19:13Z INFO OpenVPN 2.5 version: 2.5.6
2022-12-01T15:19:13Z INFO Unbound version: 1.15.0
2022-12-01T15:19:13Z INFO IPtables version: v1.8.8
2022-12-01T15:19:13Z INFO Settings summary:
├── VPN settings:
|   ├── VPN provider settings:
|   |   ├── Name: mullvad
|   |   └── Server selection settings:
|   |       ├── VPN type: openvpn
|   |       ├── Cities: malmö
|   |       ├── Owned only servers: yes
|   |       └── OpenVPN server selection settings:
|   |           └── Protocol: UDP
|   └── OpenVPN settings:
|       ├── OpenVPN version: 2.5
|       ├── User: [set]
|       ├── Password: [set]
|       ├── Network interface: tun0
|       ├── Run OpenVPN as: root
|       └── Verbosity level: 1
├── DNS settings:
|   ├── DNS server address to use: 127.0.0.1
|   ├── Keep existing nameserver(s): no
|   └── DNS over TLS settings:
|       ├── Enabled: yes
|       ├── Update period: every 24h0m0s
|       ├── Unbound settings:
|       |   ├── Authoritative servers:
|       |   |   └── cloudflare
|       |   ├── Caching: yes
|       |   ├── IPv6: no
|       |   ├── Verbosity level: 1
|       |   ├── Verbosity details level: 0
|       |   ├── Validation log level: 0
|       |   ├── System user: root
|       |   └── Allowed networks:
|       |       ├── 0.0.0.0/0
|       |       └── ::/0
|       └── DNS filtering settings:
|           ├── Block malicious: yes
|           ├── Block ads: no
|           ├── Block surveillance: no
|           └── Blocked IP networks:
|               ├── 127.0.0.1/8
|               ├── 10.0.0.0/8
|               ├── 172.16.0.0/12
|               ├── 192.168.0.0/16
|               ├── 169.254.0.0/16
|               ├── ::1/128
|               ├── fc00::/7
|               ├── fe80::/10
|               ├── ::ffff:7f00:1/104
|               ├── ::ffff:a00:0/104
|               ├── ::ffff:a9fe:0/112
|               ├── ::ffff:ac10:0/108
|               └── ::ffff:c0a8:0/112
├── Firewall settings:
|   ├── Enabled: yes
|   └── VPN input ports:
|       └── 54856
├── Log settings:
|   └── Log level: INFO
├── Health settings:
|   ├── Server listening address: 127.0.0.1:9999
|   ├── Target address: cloudflare.com:443
|   ├── Read header timeout: 100ms
|   ├── Read timeout: 500ms
|   └── VPN wait durations:
|       ├── Initial duration: 30s
|       └── Additional duration: 5s
├── Shadowsocks server settings:
|   └── Enabled: no
├── HTTP proxy settings:
|   └── Enabled: no
├── Control server settings:
|   ├── Listening address: :8000
|   └── Logging: yes
├── OS Alpine settings:
|   ├── Process UID: 1000
|   └── Process GID: 1000
├── Public IP settings:
|   ├── Fetching: every 12h0m0s
|   └── IP file path: /tmp/gluetun/ip
└── Version settings:
    └── Enabled: yes
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x0 pc=0x7bda9b]

goroutine 21 [running]:
github.com/qdm12/gluetun/internal/netlink.(*NetLink).IsIPv6Supported(0x98b8d9?)
        github.com/qdm12/gluetun/internal/netlink/ipv6.go:26 +0x1fb
main._main({0xe76b78, 0xc0001075c0}, {{0xe706c8, 0x6}, {0xe70998, 0x7}, {0xe71f20, 0x18}}, {0xc0000>
        ./main.go:297 +0x1685
main.main.func1()
        ./main.go:91 +0x150
created by main.main
        ./main.go:90 +0x51e
```

**Comment by qdm12** [comment]
@Makifun you forgot to remove the `d` option in `-itd`, so I only see the container hash logged out. Kinda weird `ip -6 route` returns ipv4 routes, although gluetun seems to work as expected for `ip -6 route`. Maybe try the last image (with commit mentioned right below :arrow_down:), it may help? :thinking: 

@Dared87 whoops good to know, hot-fixed in cc934f5c68452e6160abe244e965ea9eeeaa2854

**Comment by Makifun** [comment]
@qdm12 Excellent! It works great! Thanks! 👍 

2022-12-02T17:05:09+01:00 INFO IPv6 is not supported

**Comment by qdm12** [comment]
Awesome, I won't do a bugfix release since:
- this issue is present in all previous releases (only for Wireguard before v3.32)
- this issue is due to the docker host misbehaving (returns ipv4 routes when asked for ipv6)
- this issue doesn't affect everyone

It'll be part of release v3.33.0 which should come out soon. 🎉

**Comment by Dared87** [comment]
> @Dared87 whoops good to know, hot-fixed in [cc934f5](https://github.com/qdm12/gluetun/commit/cc934f5c68452e6160abe244e965ea9eeeaa2854)

This works now! Thanks!
